Allow setting backendbits when creating an adapter

almarklein opened the pull request:
Closes #101

almarklein commented on src/conv.rs:
```diff
@@ -179,6 +179,17 @@ pub fn map_origin3d(native: &native::WGPUOrigin3D) -> wgt::Origin3d {
+    extras: Option<&native::WGPUAdapterExtras>,
+) -> (Option<id::SurfaceId>, i32) {
+    if let Some(extras) = extras {
+        (Some(options.compatibleSurface), extras.backendBits)
```
I'm not so sure about the `Some` here. The `compatibleSurface` can be null, right?

almarklein commented on src/device.rs:
```diff
@@ -12,19 +12,38 @@ use wgc::{gfx_select, id};
 pub unsafe extern "C" fn wgpuInstanceRequestAdapter(
     _: native::WGPUInstance,
-    options: *const native::WGPURequestAdapterOptions,
+    options: &native::WGPURequestAdapterOptions,
     callback: native::WGPURequestAdapterCallback,
     userdata: *mut std::os::raw::c_void,
 ) {
-    let compatible_surface: Option<id::SurfaceId> =
-        options.as_ref().map(|options| options.compatibleSurface);
```
This shows how the `Option<id::SurfaceId>` was obtained first. But I have trouble following how this allowed the surface to be optional.

kvark replied:
this was incorrect :/

kvark commented on src/conv.rs:
```diff
@@ -179,6 +179,17 @@ pub fn map_origin3d(native: &native::WGPUOrigin3D) -> wgt::Origin3d {
+    extras: Option<&native::WGPUAdapterExtras>,
+) -> (Option<id::SurfaceId>, u32) {
+    if let Some(extras) = extras {
+        (Some(options.compatibleSurface), extras.backendBits as u32)
```
why is this always `Some`?

almarklein replied:
Because Rust thinks `options.compatibleSurface` is `SurfaceId`, not `Option<SurfaceId>`. I suspect we need to apply similar trickery as earlier to tell bindgen?

It *does* work now. I.e. I set `options.comatibleSurface` to null (when calling the C-api). So maybe the `SurfaceId` value is somehow checked down the road? Maybe it is just ignored if it does not match an existing surface?

kvark replied:
Yes, definitely, we need a similar trickery every time there is an Option<Id>

kvark commented on src/device.rs:
```diff
@@ -12,19 +12,38 @@ use wgc::{gfx_select, id};
+        WGPUSType_AdapterExtras => native::WGPUAdapterExtras)
+    );
+    // Convert backend bits. We need to initialize with *something*, thus the xor at the end.
+    let mut backend_bits = wgt::BackendBit::VULKAN;
```
why not initialize it with 0?

almarklein replied:
I don't know how. There is `None` field in `wgt::BackendBit`. And if I initialize to `0u32` I get `no implementation for u32 |= BackendBit`.

kvark replied:
`wgt::BackendBit::empty()`

kvark commented on src/device.rs:
```diff
@@ -12,19 +12,38 @@ use wgc::{gfx_select, id};
+    );
+    // Convert backend bits. We need to initialize with *something*, thus the xor at the end.
+    let mut backend_bits = wgt::BackendBit::VULKAN;
+    if given_backend_bits & native::WGPUBackendBits_Metal as u32 > 0 {
```
where are we defaulting to PRIMARY? We should be defaulting if the user doesn't pass the extra structure

almarklein replied:
Fixed.

almarklein commented:
I just saw there already is a `WGPUBackendType` enum in `webgpu.h`, but it can't be used as a bit field.

almarklein commented:
Ready for review again. I updated `build.rs` to tell Rust that `compatibleSurface` is optional.

I also changed this to make use of the existing `WGPUBackendType` enum. This results in less code, but as a consequence the user-selected backend is not a bitfield but one specific backend. Let me know if you prefer the previous solution with an actual bitfield.

The answer to that question may also depend on https://github.com/gfx-rs/wgpu-native/issues/101#issuecomment-845714542 If the automatic backend selection would be flawless, the need for overriding the behavior would be confined mostly to testing.

kvark commented:
At the end of the day, we'll follow the solution (of backend selection) that is figured out in webgpu-headers.
For now, your approach is good.

almarklein commented:
Oh, I was not expecting backend selection to be part of the webgpu API :)

kvark commented:
It's not a part of WebGPU upstream API, but it could be a part of webgpu-native.
